### ISSUE TITLE
OpenSSL 3.0.15 won't build on AIX Power machines that don't support the xxlor instruction.

### ISSUE THREAD

**Issue by MattHart-London** [openssl/openssl#25542]
Summary
OpenSSL 3.0.15 won't build on AIX Power machines that don't support xxlor instruction.
OpenSSL 3.0.14 was fine because it pre-dates [this git push](https://github.com/openssl/openssl/commit/3d3a7ecd1ae5ab08d22041f7b3b035c34f12fa02)

Details
```
perl <openssl 3.0.14 source tree>crypto/aes/asm/aesp8-ppc.pl "aix64" > /tmp/aesp8-ppc.3.0.14.s 
perl <openssl 3.0.15 source tree>crypto/aes/asm/aesp8-ppc.pl "aix64" > /tmp/aesp8-ppc.3.0.15.s
diff /tmp/aesp8-ppc.3.0.14.s /tmp/aesp8-ppc.3.0.15.s

2343   xxlor 2, 32+10, 32+10   <<----- Ouch.
2344   vsldoi  10,11,10,1
2345   xxlor 1, 32+10, 32+10   <<----- Ouch.
```
The xxlor instruction is a Power ISA (Instruction Set Architecture), specific to certain versions of the Power architecture. xxlor is supported on Power9 and Power10 processors running AIX 7.2 (with appropriate Technology Levels) and AIX 7.3.

Unfortunately we have an earlier AIX Power build machine so our OpenSSL 3.0.15 build fails at:
```
cc -Dinline= -qpic -q64 -qmaxmem=16384 -qro -qroconst -qthreaded -O -DB_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-3\"" -DMODULESDIR="\"/usr/local/lib/ossl-modules\"" -D_THREAD_SAFE -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -c -o crypto/aes/libcrypto-lib-aesp8-ppc.o crypto/aes/aesp8-ppc.s
```
With:
```
Assembler:
crypto/aes/aesp8-ppc.s: line 2343: invalid opcode or pseudo-op
crypto/aes/aesp8-ppc.s: line 2345: invalid opcode or pseudo-op
crypto/aes/aesp8-ppc.s: line 2350: invalid opcode or pseudo-op
crypto/aes/aesp8-ppc.s: line 2411: invalid opcode or pseudo-op
crypto/aes/aesp8-ppc.s: line 2412: invalid opcode or pseudo-op
crypto/aes/aesp8-ppc.s: line 2421: invalid opcode or pseudo-op
crypto/aes/aesp8-ppc.s: line 2422: invalid opcode or pseudo-op
```

Build machine
We are building on AIX 5.3
```
kylie@lucky53:~$ oslevel
5.3.0.0
pack@vaixp53:~$ ls -l /unix
lrwxrwxrwx    1 root     system           21 Jun 21 2011  /unix -> /usr/lib/boot/unix_64
kylie@lucky53:~$ prtconf
System Model: IBM,8202-E4B
Machine Serial Number: 06B1BFP
Processor Type: PowerPC_POWER7
Processor Implementation Mode: POWER 6
Processor Version: PV_6_Compat
Number Of Processors: 2
Processor Clock Speed: 3000 MHz
CPU Type: 64-bit
Kernel Type: 64-bit
LPAR Info: 2 02-vaixp53
Memory Size: 8192 MB
Good Memory Size: 8192 MB
Platform Firmware level: AL720_090
Firmware Version: IBM,AL720_090
Console Login: enable
Auto Restart: true
Full Core: false
```

**Comment by nhorman** [comment]
@dannytsen can you please look at this?  Not sure if the intent was to ignore pre-power 9 systems here, or if we need to conditionally compile the xxlor instruction only when supported.

@MattHart-London immediate workaround should be to just build with no-asm

**Comment by dannytsen** [comment]
@nhorman xxlxor is supported in ISA 2.07.  So, power8 is supported but AIX is not included.

**Comment by nhorman** [comment]
@dannytsen thank you for the quic response

To clarify, when you say "AIX is not included", is that meant to imply that the compiler on AIX for power8 is just unaware of the xxlor mnemonic?  i.e. those systems need a newer compiler to support this code?

@MattHart-London can you clarify what kind of system you are running on? You noted your system is pre-power9, but it would be good to know specifically what cpu you are using.

Trying to get to a path forward here - e.g. is the intent that power8 systems need a newer compiler version/os update to support this version of openssl?  Or are older power system users just left with no-asm builds as the only recourse?

**Comment by dannytsen** [comment]
@nhorman I am not sure what compiler AIX used.  So, a different compiler or newer one may solve the issue if xxlxor is supported.

**Comment by t8m** [comment]
It's actually this PR that backported the change to 3.0 and 3.1 https://github.com/openssl/openssl/pull/24531

**Comment by MattHart-London** [comment]
My understanding is that the xxlxor instruction was introduced in Power8 as part of its enhanced vector and SIMD (Single Instruction, Multiple Data) capabilities.

Examining the pdfs under https://wiki.raptorcs.com/wiki/Power_ISA  i see that Power ISA v 2.06+ supports xxlor, but Power ISA v 2.05 and earlier doesn't support xxlor.

My AIX 5.3 reports:
```
lsconf | grep "Processor"
Processor Type: PowerPC_POWER7
Processor Implementation Mode: POWER 6
Processor Version: PV_6_Compat
Number Of Processors: 2
```
I'll try excluding aesp8-ppc.s for my AIX build, something like:
```
"Disable AES-XTS optimization for AIX"
crypto/aes/build.info
41 -   IF[{- $target{sys_id} ne "MACOSX" -}]
41 +   IF[{- $target{sys_id} ne "AIX" && $target{sys_id} ne "MACOSX" -}]
```

Thanks for the impressive support.

**Comment by paulidale** [comment]
Should we revert the performance change?

**Comment by paulidale** [comment]
This isn't the first time this kind of thing has stung us.  I'd add:

3. Have CI tests that cover multiple revisions of a processor architecture.

Qemu tests should be able to handle this.  For PPC, we only test 64 bit LE via Qemu.  Qemu supports ppc, ppc64 and ppc64le, each with a myriad of cpu versions.  I suspect we'll need vendor or community assistance choosing a representative sample and these CIs are pretty slow but neither is a reason to not expand the selection via daily or weekly tests.

I don't recommend removing the on PR and push tests, they continue to find programming errors.  I don't recommend expanding these greatly -- these tests are expensive.

**Comment by nhorman** [comment]
I don't want to disagree with the notion of adding more ci ( who wouldn't being favor of more testing, or puppies, or feeding hungry people).  But I feel like I need to point out that power is currently a community platform.  While we certainly could add more arch variant tests for power (and other) arches, doing so adds CPU cycles to our ci set, and there is some monetary impact there, especially if we're using qemu.

It might not be a big deal, but if we were to add daily/weekly tests for something like this I would feel better doing so if 
A) IBM could donate access to the appropriate power arches to run this testing
And
B) as a community platform, we don't gate releases on passing any of these ci tests

Alternatively we could promote power to a primary platform, but that's a much larger conversation, and one I personally don't believe we have the resources to support right now

**Comment by paulidale** [comment]
I disagree that such a failure shouldn't stall a release: breaking compatibility is treated as a regression which usually gets an urgent fix.

I didn't realise GitHub had switched to CI hours rather than maximum number of simultaneous CI jobs.  Keeping costs down is always good.  Could such CIs be run on OpenSSL infrastructure instead?

**Comment by nhorman** [comment]
If we had sufficient CPU infrastructure to run it on, I don't see why not, but I think all our dedicated hardware is currently allocated.

Regarding ci blockage, I'm struggling to see m way to why we would treat power any differently from any other community platform.  We don't block release for late breaking regressions on nonstop or vms or such as release blocking as far as I can see.

I understand the prevalence of power possibly being an argument for giving it special treatment.  But rather than treat it specially, I would much rather have the debate hinge on promoting power to a primary platform and weigh it against the cost of doing so

**Comment by paulidale** [comment]
I wasn't limiting this to Power.  We've had similar problems on MIPS, ARM, S390 and probably other architectures.

**Comment by nhorman** [comment]
Thats a noble goal in principle, but not something we can undertake practically speaking ,unless the associated community maintainers can donate real hardware to allocate github runners on.

**Comment by t8m** [comment]
Please note this particular thing would NOT be find by QEMU based tests on Linux. This problem is on old AIX systems. I cannot really see how we could have such machine(s) in CI without having them been provided (and maintained) by the respective communities.

**Comment by paulidale** [comment]
QEMU emulates some very old PPC machines, I'd be extremely surprised if one of them didn't support the specific instruction used.  There are some processors emulated that date back to the early 1990s.  Justifying the inclusion of such processors in CI runs is another matter.

**Comment by t8m** [comment]
AFAIK This is a compiler thing, not a question of presence of these instructions. The code won't be used on old CPUs however the affected sources are still being built. I believe CI on contemporary Ubuntu/Debian with qemu (no matter how old CPU will be emulated) won't discover this issue on old AIX OS versions.
Yeah, an old Linux with sufficiently old gcc would probably find this on the cross-compile step, but such builder would still have to be a custom one and it is questionable how easy would it be to set it up.

**Comment by t8m** [comment]
This is unlikely to get fixed.
